File: py/data/extra.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from data import extra_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class RepeatedValueUint64:
    values: List[int] = field(default_factory=list)
    
    _proto_message: extra_pb2.RepeatedValueUint64 = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = extra_pb2.RepeatedValueUint64()

    def _populate_proto(self):
        del self._proto_message.values[:]
        for val in self.values:
            self._proto_message.values.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "RepeatedValueUint64":
        return cls(
            values=[int(val) for val in proto_message.values],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "RepeatedValueUint64":
        message = extra_pb2.RepeatedValueUint64()
        message.ParseFromString(data)
        return cls(
            values = [int(val) for val in message.values],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "RepeatedValueUint64":
        message = extra_pb2.RepeatedValueUint64()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class RepeatedValueDouble:
    values: List[float] = field(default_factory=list)
    
    _proto_message: extra_pb2.RepeatedValueDouble = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = extra_pb2.RepeatedValueDouble()

    def _populate_proto(self):
        del self._proto_message.values[:]
        for val in self.values:
            self._proto_message.values.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "RepeatedValueDouble":
        return cls(
            values=[float(val) for val in proto_message.values],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "RepeatedValueDouble":
        message = extra_pb2.RepeatedValueDouble()
        message.ParseFromString(data)
        return cls(
            values = [float(val) for val in message.values],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "RepeatedValueDouble":
        message = extra_pb2.RepeatedValueDouble()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class ValuesMap:
    timestamp: RepeatedValueUint64 = None
    valuesMap: Dict[str, RepeatedValueDouble] = field(default_factory=dict)
    
    _proto_message: extra_pb2.ValuesMap = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = extra_pb2.ValuesMap()

    def _populate_proto(self):
        if self.timestamp:
            self.timestamp._populate_proto()
            self._proto_message.timestamp.CopyFrom(self.timestamp._proto_message)
        self._proto_message.valuesMap.clear()
        for key, val in self.valuesMap.items():
            val._populate_proto()
            tmp = self._proto_message.valuesMap.setdefault(key)
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "ValuesMap":
        return cls(
            timestamp = RepeatedValueUint64._from_proto(proto_message.timestamp),
            valuesMap={key: RepeatedValueDouble._from_proto(val) for key, val in proto_message.valuesMap.items()},
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ValuesMap":
        message = extra_pb2.ValuesMap()
        message.ParseFromString(data)
        return cls(
            timestamp = (
                RepeatedValueUint64._from_proto(message.timestamp)
                if message.HasField("timestamp")
                else None
            ),
            valuesMap = {key: RepeatedValueDouble._from_proto(val) for key, val in message.valuesMap.items()},
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ValuesMap":
        message = extra_pb2.ValuesMap()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class TimeValuesPack:
    valuesPack: Dict[str, ValuesMap] = field(default_factory=dict)
    
    _proto_message: extra_pb2.TimeValuesPack = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = extra_pb2.TimeValuesPack()

    def _populate_proto(self):
        self._proto_message.valuesPack.clear()
        for key, val in self.valuesPack.items():
            val._populate_proto()
            tmp = self._proto_message.valuesPack.setdefault(key)
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "TimeValuesPack":
        return cls(
            valuesPack={key: ValuesMap._from_proto(val) for key, val in proto_message.valuesPack.items()},
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "TimeValuesPack":
        message = extra_pb2.TimeValuesPack()
        message.ParseFromString(data)
        return cls(
            valuesPack = {key: ValuesMap._from_proto(val) for key, val in message.valuesPack.items()},
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "TimeValuesPack":
        message = extra_pb2.TimeValuesPack()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/data/extra.cpp
#include "data/extra.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Data
{
RepeatedValueUint64::RepeatedValueUint64(const PbData::RepeatedValueUint64& protobuf) {
    values = {protobuf.values().begin(), protobuf.values().end()};
}

RepeatedValueUint64::operator PbData::RepeatedValueUint64() const {
    PbData::RepeatedValueUint64 ret;
    *(ret.mutable_values()) = {values.begin(), values.end()};
    return ret;
}

std::string RepeatedValueUint64::serializeAsJsonString() const {
    PbData::RepeatedValueUint64 protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string RepeatedValueUint64::serializeAsProtobufString() const {
    PbData::RepeatedValueUint64 protobuf(*this);
    return protobuf.SerializeAsString();
}

bool RepeatedValueUint64::deserializeFromJsonString(const std::string& str) {
    PbData::RepeatedValueUint64 protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool RepeatedValueUint64::deserializeFromProtobufString(const std::string& str) {
    PbData::RepeatedValueUint64 protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

RepeatedValueDouble::RepeatedValueDouble(const PbData::RepeatedValueDouble& protobuf) {
    values = {protobuf.values().begin(), protobuf.values().end()};
}

RepeatedValueDouble::operator PbData::RepeatedValueDouble() const {
    PbData::RepeatedValueDouble ret;
    *(ret.mutable_values()) = {values.begin(), values.end()};
    return ret;
}

std::string RepeatedValueDouble::serializeAsJsonString() const {
    PbData::RepeatedValueDouble protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string RepeatedValueDouble::serializeAsProtobufString() const {
    PbData::RepeatedValueDouble protobuf(*this);
    return protobuf.SerializeAsString();
}

bool RepeatedValueDouble::deserializeFromJsonString(const std::string& str) {
    PbData::RepeatedValueDouble protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool RepeatedValueDouble::deserializeFromProtobufString(const std::string& str) {
    PbData::RepeatedValueDouble protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

ValuesMap::ValuesMap(const PbData::ValuesMap& protobuf) {
    timestamp = protobuf.timestamp();
    valuesMap = {protobuf.valuesmap().begin(), protobuf.valuesmap().end()};
}

ValuesMap::operator PbData::ValuesMap() const {
    PbData::ValuesMap ret;
    *(ret.mutable_timestamp()) = timestamp;
    *(ret.mutable_valuesmap()) = {valuesMap.begin(), valuesMap.end()};
    return ret;
}

std::string ValuesMap::serializeAsJsonString() const {
    PbData::ValuesMap protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ValuesMap::serializeAsProtobufString() const {
    PbData::ValuesMap protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ValuesMap::deserializeFromJsonString(const std::string& str) {
    PbData::ValuesMap protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ValuesMap::deserializeFromProtobufString(const std::string& str) {
    PbData::ValuesMap protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

TimeValuesPack::TimeValuesPack(const PbData::TimeValuesPack& protobuf) {
    valuesPack = {protobuf.valuespack().begin(), protobuf.valuespack().end()};
}

TimeValuesPack::operator PbData::TimeValuesPack() const {
    PbData::TimeValuesPack ret;
    *(ret.mutable_valuespack()) = {valuesPack.begin(), valuesPack.end()};
    return ret;
}

std::string TimeValuesPack::serializeAsJsonString() const {
    PbData::TimeValuesPack protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string TimeValuesPack::serializeAsProtobufString() const {
    PbData::TimeValuesPack protobuf(*this);
    return protobuf.SerializeAsString();
}

bool TimeValuesPack::deserializeFromJsonString(const std::string& str) {
    PbData::TimeValuesPack protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool TimeValuesPack::deserializeFromProtobufString(const std::string& str) {
    PbData::TimeValuesPack protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}